gap/matrix/slconstr.gi

```diff
@@ -2482,16 +2482,21 @@ FindHomMethodsMatrix.NaturalSL := function(ri,G)
     fi;
 
 
+  # Try the recognition method
   n:=ri!.dimension;
   q:=Size(ri!.field);
   p:=Factors(q)[1];
   e:=LogInt(q,p);
   grpmem := GroupWithMemory(G);
-  repeat
-      data:=SLCR.SLDataStructure(grpmem,p,e,n);
-  until data <> fail;
-  genlist:=List(data.gens,x->x[1]);
-  data.slpnice:=SLPOfElms(genlist);
+  # TODO find a reasonable number of tries
+  for i in [1 .. 10] do
+    data := SLCR.SLDataStructure(grpmem, p, e, n);
+    if data <> fail then break; fi;
+  od;
+  if data <> fail then
+    genlist := List(data.gens, x->x[1]);
+    data.slpnice := SLPOfElms(genlist);
+  fi;
 
   # Now clean out the memory everywhere in data:
   for i in [1..Length(data.gens)] do
```

Comment:
Won't this run into an error if `data = fail`? I think the `data = fail` test needs to be moved before this.

Reply:
Done.